could not read file: jar:file

looks like we have this bug

http://groovyjavathoughts.blogspot.co.uk/2015/01/why-am-i-getting-warning-arguments-left.html

https://github.com/ensime/ensime-server/blob/2.0/core/src/main/scala/org/ensime/indexer/SearchService.scala#L337

which means there is probably another bug hiding under this.

```
org.apache.commons.vfs2.FileSystemException: Could not read file "jar:file:///C:/FOO/scala-2.11/last/BAR.jar!/foo/Bar.class".
	at org.apache.commons.vfs2.provider.AbstractFileObject.getInputStream(AbstractFileObject.java:1317)
	at org.apache.commons.vfs2.provider.DefaultFileContent.getInputStream(DefaultFileContent.java:411)
	at org.ensime.indexer.ClassfileIndexer$class.indexClassfile(ClassfileIndexer.scala:24)
	at org.ensime.indexer.SearchService.indexClassfile(SearchService.scala:28)
	at org.ensime.indexer.SearchService.org$ensime$indexer$SearchService$$extractSymbols(SearchService.scala:212)
	at org.ensime.indexer.SearchService$$anonfun$extractSymbolsFromClassOrJar$1$$anonfun$apply$7$$anonfun$apply$8.apply(SearchService.scala:196)
	at org.ensime.indexer.SearchService$$anonfun$extractSymbolsFromClassOrJar$1$$anonfun$apply$7$$anonfun$apply$8.apply(SearchService.scala:196)
	at scala.collection.immutable.List.flatMap(List.scala:327)
	at org.ensime.indexer.SearchService$$anonfun$extractSymbolsFromClassOrJar$1$$anonfun$apply$7.apply(SearchService.scala:196)
	at org.ensime.indexer.SearchService$$anonfun$extractSymbolsFromClassOrJar$1$$anonfun$apply$7.apply(SearchService.scala:183)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
	at scala.concurrent.package$.blocking(package.scala:123)
	at org.ensime.indexer.SearchService$$anonfun$extractSymbolsFromClassOrJar$1.apply(SearchService.scala:183)
	at org.ensime.indexer.SearchService$$anonfun$extractSymbolsFromClassOrJar$1.apply(SearchService.scala:183)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: zip file closed
	at java.util.zip.ZipFile.ensureOpen(ZipFile.java:634)
	at java.util.zip.ZipFile.getInputStream(ZipFile.java:347)
	at java.util.jar.JarFile.getInputStream(JarFile.java:412)
	at org.apache.commons.vfs2.provider.zip.ZipFileObject.doGetInputStream(ZipFileObject.java:174)
	at org.apache.commons.vfs2.provider.AbstractFileObject.getInputStream(AbstractFileObject.java:1301)
	... 19 common frames omitted

```

backporting the fix for that small problem I can reproduce locally by typing `testingSimpleJar/clean` and `testingSimpleJar/packageBin` in quick succession (or just a `touch testing/simpleJar/target/scala-2.11/testingsimplejar_2.11-1.0.0-SNAPSHOT.jar`)

```
org.apache.commons.vfs2.FileSystemException: Could not read file "jar:file:///home/fommil/Projects/ensime-server/testing/simpleJar/target/scala-2.11/testingsimplejar_2.11-1.0.0-SNAPSHOT.jar!/baz/Foo.class".
	at org.apache.commons.vfs2.provider.AbstractFileObject.getInputStream(AbstractFileObject.java:1317)
	at org.apache.commons.vfs2.provider.DefaultFileContent.getInputStream(DefaultFileContent.java:411)
	at org.ensime.indexer.ClassfileIndexer$class.indexClassfile(ClassfileIndexer.scala:24)
	at org.ensime.indexer.SearchService.indexClassfile(SearchService.scala:28)
	at org.ensime.indexer.SearchService.org$ensime$indexer$SearchService$$extractSymbols(SearchService.scala:212)
	at org.ensime.indexer.SearchService$$anonfun$extractSymbolsFromClassOrJar$1$$anonfun$apply$7$$anonfun$apply$8.apply(SearchService.scala:196)
	at org.ensime.indexer.SearchService$$anonfun$extractSymbolsFromClassOrJar$1$$anonfun$apply$7$$anonfun$apply$8.apply(SearchService.scala:196)
	at scala.collection.immutable.List.flatMap(List.scala:327)
	at org.ensime.indexer.SearchService$$anonfun$extractSymbolsFromClassOrJar$1$$anonfun$apply$7.apply(SearchService.scala:196)
	at org.ensime.indexer.SearchService$$anonfun$extractSymbolsFromClassOrJar$1$$anonfun$apply$7.apply(SearchService.scala:183)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
	at scala.concurrent.package$.blocking(package.scala:123)
	at org.ensime.indexer.SearchService$$anonfun$extractSymbolsFromClassOrJar$1.apply(SearchService.scala:183)
	at org.ensime.indexer.SearchService$$anonfun$extractSymbolsFromClassOrJar$1.apply(SearchService.scala:183)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: zip file closed
	at java.util.zip.ZipFile.ensureOpen(ZipFile.java:634)
	at java.util.zip.ZipFile.getInputStream(ZipFile.java:347)
	at java.util.jar.JarFile.getInputStream(JarFile.java:412)
	at org.apache.commons.vfs2.provider.zip.ZipFileObject.doGetInputStream(ZipFileObject.java:174)
	at org.apache.commons.vfs2.provider.AbstractFileObject.getInputStream(AbstractFileObject.java:1301)
	... 19 common frames omitted

```

--- a/core/src/main/scala/org/ensime/indexer/database/DatabaseService.scala
+++ b/core/src/main/scala/org/ensime/indexer/database/DatabaseService.scala
@@ -125,7 +125,7 @@ object DatabaseService {
   object FileCheck extends ((Option[Int], String, Timestamp) => FileCheck) {
     def apply(f: FileObject): FileCheck = {
       val name = f.getName.getURI
-      val ts = new Timestamp(f.getContent.getLastModifiedTime)
+      val ts = new Timestamp(if (!f.exists()) -1L else f.getContent.getLastModifiedTime)
       FileCheck(None, name, ts)
     }
   }

--- a/util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala
+++ b/util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala
@@ -101,6 +101,7 @@ object `package` {
         zip.asInstanceOf[ZipFile].close()
 
       vfs.getFilesCache.clear(fs)
+      vfs.closeFileSystem(fs)
     }
   }
 }

